Jetpack Settings: Add jetpack module activation reducers, actions types, action creators and selectors

oskosk opened the pull request:
Part of #9171 

This PR is a kick-off for implementing Jetpack Settings and Module related operation in the redux state tree.

#### Changes proposed in this Pull Request:

* Implements `client/state/jetpack-settings/reducer.js`
  * Introduces a `jetpackSettings` leaf on the redux state tree.
  * Enables this reducer in `client/state/index.js`.
* Implements `client/state/jetpack-settings/modules/reducer.js`. _The `jetpack-settings/reducer` combines this reducer_. 
  * Introduces the `jetpackSettings.jetpackModules` leaf in the state tree.
* Introduces an action creator `activateModule` for activating a jetpack module.
  * It's based on a non-existant `lib/wp.undocumented().activateJetpackModule()` method. _The action tests mock this method for passing_.
* Introduces the selectors `isActivatingModule()` and `isModuleActivated` for querying the state.

#### Testing instructions

1. Run 
    ```
    $ npm run test-client client/state/jetpack-settings/
    ```
1. Expect the tests to pass.

oskosk commented:
cc @tyxla @johnHackworth

tyxla commented on client/state/jetpack-settings/modules/reducer.js:
```diff
@@ -0,0 +1,80 @@
+ * @param {Object} action - action
+ * @return {Object} updated state
+ */
+export const requests = ( state = initialRequestsState, action ) => {
```
Here we'll also have to track modules by `siteId`. Using the above suggestions, I'd change it to:

```
export const requests = ( state = initialRequestsState, { type, siteId, module } ) => {
	switch ( type ) {
		case JETPACK_MODULE_ACTIVATE:
			return merge( {}, state, {
				[ siteId ]: {
					[ module ]: true
				}
			} );
		case JETPACK_MODULE_ACTIVATE_FAIL:
		case JETPACK_MODULE_ACTIVATE_SUCCESS:
			return merge( {}, state, {
				[ siteId ]: {
					[ module ]: false
				}
			} );
		// return initial state when serializing/deserializing
		case SERIALIZE:
		case DESERIALIZE:
			return initialRequestsState;
		default:
			return state;
	}
};
```

tyxla commented on client/state/jetpack-settings/modules/actions.js:
```diff
@@ -0,0 +1,41 @@
+} from 'state/action-types';
+import wp from 'lib/wp';
+
+export const activateModule = ( module_slug ) => {
```
Let's use the recommended camel case for `module_slug` and change it to `moduleSlug` everywhere.

oskosk replied:
nice, I was waiting for review on how to handle this context.

oskosk replied:
alright about the camelcase!

tyxla commented on client/state/jetpack-settings/modules/test/actions.js:
```diff
@@ -0,0 +1,48 @@
+		useSandbox( newSandbox => sandbox = newSandbox );
+
+		it( 'should dispatch complete success when API activates a module', () => {
+			sandbox.stub( wp, 'undocumented', () => ( {
```
Once we switch the `activateJetpackModule` to return a `Promise`, lets also use `nock` to emulate the http requests.

tyxla commented on client/state/jetpack-settings/modules/test/actions.js:
```diff
@@ -0,0 +1,48 @@
+
+			return result.then( () => {
+				expect( spy ).to.have.been.calledWith( {
+					type: JETPACK_MODULE_ACTIVATE_SUCCESS,
```
The new `siteId` action prop will have to be addressed in the tests as well.

tyxla commented on client/state/jetpack-settings/modules/test/actions.js:
```diff
@@ -0,0 +1,48 @@
+		spy.reset();
+	} );
+
+	describe( '#activateJetpackModule', () => {
```
We should add tests for all dispatched actions.

oskosk commented on client/state/jetpack-settings/modules/actions.js:
```diff
@@ -0,0 +1,41 @@
+			type: JETPACK_MODULE_ACTIVATE,
+			module: module_slug
+		} );
+		return new Promise( ( resolve, reject ) => {
```
Ah  yes. It's just that I thought that wpcomjs was generally used with the callback mechanism. In fact I rewrote that yesterday from looking as you propose to how it is now

johnHackworth replied:
https://github.com/Automattic/wp-calypso/blob/master/client/lib/wpcom-undocumented/lib/undocumented.js#L81

Why are we defining a new 'activateJetpackModule' instead of using the existing 'jetpackModuleActivate' ?

tyxla replied:
Ha, I didn't know that method exists already. Good find! Let's use it here, it appears to be exactly what we need.

oskosk commented on client/state/jetpack-settings/modules/actions.js:
```diff
@@ -0,0 +1,41 @@
+			module: module_slug
+		} );
+		return new Promise( ( resolve, reject ) => {
+			wp.undocumented().activateJetpackModule( module_slug, ( error, data ) => {
```
perfect

oskosk commented on client/state/jetpack-settings/modules/reducer.js:
```diff
@@ -0,0 +1,80 @@
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import assign from 'lodash/assign';
```
Are we safe to use Object.assign now (regarding browsers we support) ? cool

johnHackworth replied:
we can in calypso ... we just support latest version plus one :)

oskosk commented on client/state/jetpack-settings/modules/actions.js:
```diff
@@ -0,0 +1,41 @@
+				type: JETPACK_MODULE_ACTIVATE_FAIL,
+				module: module_slug,
+				success: false,
+				error: error.message || i18n.translate( 'There was a problem activating the module.' )
```
I didn't quite catch this.

tyxla replied:
I'm just saying i'm not sure this is the best place to have a default error message set. I'd hazard a guess that it would be better if we stick just with `error.message` here. Main reason for my assumption: we might want to know if there was an error with no specific error message, because then we can define a custom error message inline in the component.

oskosk commented:
Thank you @tyxla . I've just updated this PR, so some comments may have disappeared . Here's the summary of the changes made as per your observations:

- [x] Move wpcom.undocumented() to wpcom.undecomunted.site() 
- [x] add siteId parameter in activateModule action creator
- [x] add siteId parameter in call to wp.undocumented().site()
- [x] add siteId parameter in dispatched action s
- [x] add siteId parameter in action tests
- [x] Index modules and requests by siteid in redux state tree structure
- [x] Update reducers to take in account siteId indexation
- [x] Use Camelcase for moduleSlug variable
- [x] Rename  the action typeJETPACK_MODULE_ACTIVATE_FAIL to JETPACK_MODULE_ACTIVATE_FAILURE
- [ ] Tests for all dispatched actions in the items reducer tests
- [ ] Use nock to emulate http requests
- [ ] Add isModuleActivated and isActivatingModule selectors
- [x] Update code to assume that wp.undocumented().site().activateModule() returns a promise.

oskosk commented:
@tyxla I've implemented the selectors now. 

Also I've replaced usage of the to-be-defined `wp.undocumented().site().activateJetpackModule` for the already existing `wp.undocumented().jetpackModulesActivate()` as @johnHackworth proposed, which seems to work.

tyxla commented on client/state/jetpack-settings/modules/reducer.js:
```diff
@@ -0,0 +1,88 @@
+	switch ( type ) {
+		case JETPACK_MODULE_ACTIVATE:
+			return merge( {}, state, {
+				activating: {
```
I imagined the state tree structure to be indexed by site, then by module, like this:

```
[ siteId ]: {
	[ moduleSlug ]: {
		activating: true
	}
}
```

This way we will not have to duplicate this structure for other module data like `deactivating`, or `enabled` or anything else we might want to store. Because if we keep what you're suggesting here, it would eventually become like:

```
activating: {
	[ siteId ]: {
		[ moduleSlug ]: true
	}
},
deactivating: {
	[ siteId ]: {
		[ moduleSlug ]: false
	}
}
```

while with what I'm suggesting it would avoid the `siteId` and `moduleSlug` duplication:

```
[ siteId ]: {
	[ moduleSlug ]: {
		activating: true,
		deactivating: false
	}
}
```

oskosk replied:
You're right! I'm updating the code now, thanks!

tyxla commented on client/state/jetpack-settings/modules/test/actions.js:
```diff
@@ -0,0 +1,88 @@
+			} );
+		} );
+	} );
+	describe( '#activateJetpackModuleSuccess', () => {
```
I'd wrap all tests for an action within the `describe()` call of the action, instead of just creating another for each case. So the structure of the tests would become like:

- actions
  - activateJetpackModule
    - success
      - `JETPACK_MODULE_ACTIVATE`
      - `JETPACK_MODULE_ACTIVATE_SUCCESS`
    - failure
      - `JETPACK_MODULE_ACTIVATE_FAILURE`

or

- actions
  - activateJetpackModule
    - `JETPACK_MODULE_ACTIVATE`
    - success
      - `JETPACK_MODULE_ACTIVATE_SUCCESS`
    - failure
      - `JETPACK_MODULE_ACTIVATE_FAILURE`

tyxla commented on client/state/jetpack-settings/modules/test/reducer.js:
```diff
@@ -0,0 +1,89 @@
+	requests as REQUESTS_FIXTURE
+} from './fixture';
+
+describe( 'items reducer', () => {
```
I'd prefer wrapping all reducer tests in a main `describe( 'reducer', () => {` call.

tyxla commented on client/state/jetpack-settings/modules/test/selectors.js:
```diff
@@ -0,0 +1,48 @@
+	requests as REQUESTS_FIXTURE
+} from './fixture';
+
+describe( 'requests selectors', () => {
```
Some selectors might not be based on requests / items. So it would be best to avoid this grouping. Instead, let's add a single `describe( 'selectors', () => {` that wraps all selectors.

tyxla commented:
I've posted some very minor nitpicking comments, it should be good to go after addressing.

tyxla commented on client/state/jetpack-settings/modules/selectors.js:
```diff
@@ -0,0 +1,27 @@
+import get from 'lodash/get';
+
+/**
+ * Returns true if the module is activated
```
Let's change `activated` in docs to `active` - for consistency.

oskosk commented:
Thanks @tyxla . I've just updated the code according to your recommendations. Can you check if the wrapping for the action tests is the way you mentioned now ?

tyxla commented on client/state/jetpack-settings/modules/test/reducer.js:
```diff
@@ -0,0 +1,91 @@
+			expect( state ).to.eql( {} );
+		} );
+
+		describe( '#modulsActivation', () => {
```
In the future there might be some actions that can't be classified as just one type of actions (modules activation). In that case we'd have to refactor these `describe` calls I'd remove that `describe`. But at this time it makes sense, so that's your call, really. 

Ah, and there's a typo - `modulsActivation` - `s` should be `e`.

tyxla commented:
LGTM 🚢
